working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+/// corresponding [resource].
+class Wrapper {
+  /// Native resource held by the wrapper. Lifetime
+  /// of the wrapper determines lifetime of the [_resource].
+  final ffi.Pointer<ffi.Void> _resource;
+
+  void method() {
+    // [this] might be reclaimed by GC after evaluating
+    // [this.resource] because it is never used again
+    // and not kept live
+    useResource(this._resource, causeGC());
+  }
+}
+```
+
+In this example if finalization can occur synchronously when `causeGC` is called
+`this._resource` might be released prematurely because no references to `this`
+remain alive after `this._resource` is evaluated and saved as an outgoing
+argument. This means that a garbage collection which occurs within `causeGC()`
+might reclaim a `Wrapper` object which was previous pointed by `this` and
+consequently invoke finalization callback associated with it, which in turn
+would release the native resource.
+
+## Proposal
+
+We split the proposed finalization API into two parts:
+
+- `FinalizationRegistry` which:
+  - does not provide strong guarantees around promptness of finalization;
+  - does not impose any restrictions on objects, you could associate a
+    finalization action with;
+  - does not impose any restrictions on finalization actions, but invokes them
+    asynchronously;
+  - is implementable on the Web.
+- `NativeFinalizationRegistry` which:
+  - provides stronger guarantees around promptness of finalization;
+  - guarantees that finalization actions are invoked when isolate is shutting
+  down;
+  - is limited to objects which implement `Finalizable` interface;
+  - imposes restrictions on finalization actions to allow calling these actions
+  outside of Dart universe.
+
+The following classes are added to `dart:core` library
+
+```dart
+/// A register of objects which may invoke a callback when those objects
+/// become inaccessible.
+///
+/// The register allows objects to be registered,
+/// and when those objects become inaccessible to the program.
+/// the callback passed to the register's constructor *may* be called
+/// with the registration token associated with the object.
+///
+/// No promises are made that the callback will ever be called,
+/// only that *if* it is called with a finalization token as argument,
+/// at least one object registered in the registry with that finalization token
+/// is no longer accessible to the program.
+///
+/// If the same object is registered in multiple finalization registries,
+/// or registered multiple times in a single registry,
+/// and the object becomes inaccessible to the program,
+/// then any number of those registrations may trigger their associated
+/// callback. It will not necessarily be all or none of them.
+///
+/// Finalization callbacks will happen as *events*, not during execution of
+/// other code and not as a microtask, but as high-level events similar to
+/// timer events.
+abstract class FinalizationRegistry<F extends Object, FT> {
```

Comment:
Drop the `F` type parameter. It's not needed for anything.

Reply:
Alternative is to use `covariant` on the `register` argument - wanted to avoid that :)